OpenSim/Simulation/Model/OffsetFrame.cpp

```diff
@@ -0,0 +1,209 @@
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+//=============================================================================
+// INCLUDES
+//=============================================================================
+#include "OffsetFrame.h"
+#include "PhysicalFrame.h"
+
+//=============================================================================
+// STATICS
+//=============================================================================
+using namespace std;
+
+namespace OpenSim {
+
+//=============================================================================
+// CONSTRUCTOR(S)
+//=============================================================================
+//_____________________________________________________________________________
+/**
+ * Default constructor.
+ */
+template <class C>
+OffsetFrame<C>::OffsetFrame() : C()
+{
+    setNull();
+    this->constructInfrastructure();
+}
+
+//_____________________________________________________________________________
+/**
+ * Constructors.
+ */
+template <class C>
+OffsetFrame<C>::OffsetFrame(const C& parent,
+    const SimTK::Transform& offset) : C()
+{
+    setNull();
+    this->constructInfrastructure();
+    setParentFrame(parent);
+    setOffsetTransform(offset);
+}
+//_____________________________________________________________________________
+/**
+* Set a null frame as Identity rotation, 0 translation
+*/
+template <class C>
+void OffsetFrame<C>::setNull()
+{
+    _offsetTransform.setToNaN();
+    this->setAuthors("Matt DeMers, Ajay Seth");
+}
+//_____________________________________________________________________________
+/**
+* construct properties
+*/
+template <class C>
+void OffsetFrame<C>::constructProperties()
+{
+    SimTK::Vec3 zero(0.0, 0.0, 0.0);
+    constructProperty_translation(zero);
+    constructProperty_orientation(zero);
+    // transform at default
+}
+
+template <class C>
+void OffsetFrame<C>::constructConnectors()
+{
+    this->template constructConnector<C>("parent");
+}
+
+//=============================================================================
+// FRAME COMPUTATIONS
+//=============================================================================
+//_____________________________________________________________________________
+/*
+ * Implementation of Frame interface by OffsetFrame.
+ *
+ */
+template <class C>
+const SimTK::Transform& OffsetFrame<C>::
+    calcGroundTransform(const SimTK::State& s) const
+{
+    if (!this->isCacheVariableValid(s, "ground_transform")){
```

Comment:
Ouch -- that looks expensive with the string lookup. I hope it doesn't get called much!